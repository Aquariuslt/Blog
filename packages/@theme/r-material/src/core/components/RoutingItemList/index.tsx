import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { makeStyles, StyleRules, Theme, useTheme } from '@material-ui/core/styles';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import Collapse from '@material-ui/core/Collapse';

import { NavMenu } from '../../stores/navigation.store';
import BundledIcon from '../../components/BundledIcon';
import RoutingItem from '../../components/RoutingItem';

type RoutingItemListProps = NavMenu;

const useStyles = makeStyles(
  (theme: Theme): StyleRules => ({
    root: {
      width: '100%',
      maxWidth: 360,
      backgroundColor: theme.palette.background.paper,
      paddingTop: 0,
      paddingBottom: 0
    }
  })
);

const RoutingItemList: React.ComponentType<RoutingItemListProps> = (props: RoutingItemListProps): JSX.Element => {
  const theme: Theme = useTheme();
  const classes = useStyles(theme);

  const [open, setOpen] = React.useState(false);
  const [t] = useTranslation();

  function toggleOpen(): void {
    setOpen(!open);
  }

  return (
    <List component="nav" className={classes.root}>
      <ListItem component="div" button onClick={toggleOpen} aria-label="Toggle Menu">
        <ListItemIcon>
          <BundledIcon type={props.icon || 'menu'} />
        </ListItemIcon>
        <ListItemText primary={t(props.name)} />
        {open ? <BundledIcon type="expandLess" /> : <BundledIcon type="expandMore" />}
      </ListItem>
      <Collapse in={open} timeout="auto" unmountOnExit>
        <List component="div" disablePadding>
          {props.child.map(
            (childProps, index): JSX.Element => (
              <RoutingItem key={index} {...childProps} />
            )
          )}
        </List>
      </Collapse>
    </List>
  );
};

export default RoutingItemList;
