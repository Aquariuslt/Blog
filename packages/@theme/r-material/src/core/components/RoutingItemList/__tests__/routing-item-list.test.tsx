import * as React from 'react';
import * as Enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });

import RoutingItemList from '../';

describe('@theme/r-material: core/components: RoutingItemList', (): void => {
  it('# should mount routing item component without child correctly', (): void => {
    const navMenuWoChild = {
      name: 'mockName',
      link: '/to/somewhere',
      icon: 'menu',
      priority: 3,
      child: []
    };

    const wrapper = Enzyme.shallow(<RoutingItemList {...navMenuWoChild} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('# should mount routing item component with child correctly', (): void => {
    const navMenuWChild = {
      name: 'mockName',
      link: '/to/somewhere',
      icon: 'menu',
      priority: 3,
      child: [
        {
          name: 'mock-sub-name',
          link: '/to/some-else'
        },
        {
          name: 'mock-other-sub-name',
          link: '/to/bar'
        }
      ]
    };

    const wrapper = Enzyme.shallow(<RoutingItemList {...navMenuWChild} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('# should expand item list with child correctly', (): void => {
    const navMenuWChild = {
      name: 'mockName',
      link: '/to/somewhere',
      icon: 'menu',
      priority: 3,
      child: [
        {
          name: 'mock-sub-name',
          link: '/to/some-else'
        },
        {
          name: 'mock-other-sub-name',
          link: '/to/bar'
        }
      ]
    };

    const wrapper = Enzyme.shallow(<RoutingItemList {...navMenuWChild} />);
    wrapper.find('[aria-label="Toggle Menu"]').simulate('click');
  });
});
