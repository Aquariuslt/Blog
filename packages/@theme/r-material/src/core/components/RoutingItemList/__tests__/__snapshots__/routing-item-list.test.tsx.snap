// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@theme/r-material: core/components: RoutingItemList # should mount routing item component with child correctly 1`] = `
<WithStyles(ForwardRef(List))
  className="makeStyles-root-1"
  component="nav"
>
  <WithStyles(ForwardRef(ListItem))
    aria-label="Toggle Menu"
    button={true}
    component="div"
    onClick={[Function]}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <BundledIcon
        type="menu"
      />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="mockName"
    />
    <BundledIcon
      type="expandMore"
    />
  </WithStyles(ForwardRef(ListItem))>
  <WithStyles(ForwardRef(Collapse))
    in={false}
    timeout="auto"
    unmountOnExit={true}
  >
    <WithStyles(ForwardRef(List))
      component="div"
      disablePadding={true}
    >
      <RoutingItem
        key="0"
        link="/to/some-else"
        name="mock-sub-name"
      />
      <RoutingItem
        key="1"
        link="/to/bar"
        name="mock-other-sub-name"
      />
    </WithStyles(ForwardRef(List))>
  </WithStyles(ForwardRef(Collapse))>
</WithStyles(ForwardRef(List))>
`;

exports[`@theme/r-material: core/components: RoutingItemList # should mount routing item component without child correctly 1`] = `
<WithStyles(ForwardRef(List))
  className="makeStyles-root-1"
  component="nav"
>
  <WithStyles(ForwardRef(ListItem))
    aria-label="Toggle Menu"
    button={true}
    component="div"
    onClick={[Function]}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <BundledIcon
        type="menu"
      />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="mockName"
    />
    <BundledIcon
      type="expandMore"
    />
  </WithStyles(ForwardRef(ListItem))>
  <WithStyles(ForwardRef(Collapse))
    in={false}
    timeout="auto"
    unmountOnExit={true}
  >
    <WithStyles(ForwardRef(List))
      component="div"
      disablePadding={true}
    />
  </WithStyles(ForwardRef(Collapse))>
</WithStyles(ForwardRef(List))>
`;

exports[`@theme/r-material: core/components: RoutingItemList # should mount routing item component without icon correctly 1`] = `
<WithStyles(ForwardRef(List))
  className="makeStyles-root-1"
  component="nav"
>
  <WithStyles(ForwardRef(ListItem))
    aria-label="Toggle Menu"
    button={true}
    component="div"
    onClick={[Function]}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <BundledIcon
        type="menu"
      />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="mockName"
    />
    <BundledIcon
      type="expandMore"
    />
  </WithStyles(ForwardRef(ListItem))>
  <WithStyles(ForwardRef(Collapse))
    in={false}
    timeout="auto"
    unmountOnExit={true}
  >
    <WithStyles(ForwardRef(List))
      component="div"
      disablePadding={true}
    />
  </WithStyles(ForwardRef(Collapse))>
</WithStyles(ForwardRef(List))>
`;
